core/foundation/inc/ROOT/TypeTraits.hxx

```diff
@@ -0,0 +1,203 @@
+   static constexpr bool Test(...)
+   {
+      return false;
+   }
+
+   static constexpr bool value = Test<Test_t>(nullptr);
+};
+
+/// Check whether a histogram type is a classic or v7 histogram.
+template <typename T>
+struct IsV7Hist : public std::false_type {
+   static_assert(std::is_base_of<TH1, T>::value, "not implemented for this type");
+};
+
+template <int D, typename P, template <int, typename, template <typename> class> class... S>
+struct IsV7Hist<ROOT::Experimental::THist<D, P, S...>> : public std::true_type {
+};
+
+/// Lightweight storage for a collection of types.
+/// Differently from std::tuple, no instantiation of objects of stored types is performed
+template <typename... Types>
+struct TypeList {
+   static constexpr std::size_t list_size = sizeof...(Types);
+};
+
+/// Extract types from the signature of a callable object.
+template <typename T>
+struct CallableTraits {
+   using arg_types = typename CallableTraits<decltype(&T::operator())>::arg_types;
+   using arg_types_nodecay = typename CallableTraits<decltype(&T::operator())>::arg_types_nodecay;
+   using ret_type = typename CallableTraits<decltype(&T::operator())>::ret_type;
+};
+
+// lambdas and std::function
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...) const> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// mutable lambdas and functor classes
+template <typename R, typename T, typename... Args>
+struct CallableTraits<R (T::*)(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// function pointers
+template <typename R, typename... Args>
+struct CallableTraits<R (*)(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// free functions
+template <typename R, typename... Args>
+struct CallableTraits<R(Args...)> {
+   using arg_types = TypeList<typename std::decay<Args>::type...>;
+   using arg_types_nodecay = TypeList<Args...>;
+   using ret_type = R;
+};
+
+// Return first of a variadic list of types.
+template <typename T, typename... Rest>
+struct TakeFirst {
+   using type = T;
+};
+
+template <typename... Types>
+using TakeFirst_t = typename TakeFirst<Types...>::type;
+
+// Remove first type from a variadic list of types, return a TypeList containing the rest.
+// e.g. RemoveFirst_t<A,B,C> is TypeList<B,C>
+template <typename T, typename... Rest>
+struct RemoveFirst {
+   using type = TypeList<Rest...>;
+};
+
+template <typename... Args>
+using RemoveFirst_t = typename RemoveFirst<Args...>::type;
+
+/// Return first of possibly many template parameters.
+/// For non-template types, the result is the type itself.
+/// e.g. TakeFirstParameter<U<A,B>> is A
+///      TakeFirstParameter<T> is T
```

Comment:
That is a very ... rare ... behavior. Providing this as a general tool might see misuses / misunderstanding of the interface. I'd naively expect that `TakeFirstParameter<T>` is either empty or ill-formed, instead of mixing template levels... Do you expect generic usage of this? (Do you really know you need this behavior in your use case? I didn't look at it, it seems surprising...)